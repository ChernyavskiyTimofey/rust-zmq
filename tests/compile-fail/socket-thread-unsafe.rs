use std::thread;

macro_rules! t {
    ($e:expr) => (
        $e.unwrap_or_else(|e| { panic!("{} failed with {:?}", stringify!($e), e) })
    )
}

fn main() {
    let mut context = zmq2::Context::new();
    let socket = t!(context.socket(zmq2::REP));
    let s = &socket;
    let t = thread::spawn(move || {
        t!(s.bind("tcp://127.0.0.1:12345"))
    });
    socket.send("ABC", 0);
    t.join().unwrap();
}
