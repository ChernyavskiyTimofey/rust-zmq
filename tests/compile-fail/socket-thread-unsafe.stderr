error[E0277]: `*mut std::ffi::c_void` cannot be shared between threads safely
   --> $DIR/socket-thread-unsafe.rs:13:13
    |
13  |     let t = thread::spawn(move || {
    |             ^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be shared between threads safely
    |
    = help: within `zmq2::Socket`, the trait `std::marker::Sync` is not implemented for `*mut std::ffi::c_void`
    = note: required because it appears within the type `zmq2::Socket`
    = note: required because of the requirements on the impl of `std::marker::Send` for `&zmq2::Socket`
    = note: required because it appears within the type `[closure@$DIR/tests/compile-fail/socket-thread-unsafe.rs:13:27: 15:6 s:&zmq2::Socket]`
